### ISSUE TITLE
Update Firefox to ESR31

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/mozilla-os2#100]
We need to update our Firefox codebase to ESR31. This will include restoring OS/2 parts removed between ESR24 and ESR31 back in business.

**Comment by dmik** [comment]
According to http://hg.mozilla.org/releases/mozilla-esr31/, the latest ESR31 release is 31.6.0, FIREFOX_31_6_0esr_RELEASE, released Wed, 25 Mar 2015 23:30:15 -0400. This one I am about to pull into our repository.

**Comment by dmik** [comment]
I've committed `FIREFOX_31_6_0esr_RELEASE` to `vendor-esr` in 1fd092e86ee88b0ed46470ca969eb6b140e2ea59. Things I briefly noticed:

- OS/2 code is removed indeed.
- Many tests have been replaced / reorganized (the majority of changes seems to be here).
- Some parts got huge update (gfx, js-jit, stlport).

Merging these changes into `master` will surely not be very easy.

Note that I've also started off https://github.com/bitwiseworks/mozilla-os2/wiki/Developers to put developer-related note there and described our repository structure so that it's easier to maintain it in the future.

**Comment by dryeo** [comment]
On 04/07/15 04:48 PM, Dmitriy Kuminov wrote:
> OS/2 code is removed indeed.

There's still some in nsprpub, nss and ldap/sdks (used by TB/SM). 
Comm-central has also been stripped of OS/2 code.

**Comment by dmik** [comment]
I guess that they consider nsprpub a separate project so the cleanup doesn't touch it.

**Comment by dmik** [comment]
BTW, it seems that most work of removing the OS/2 code was done here: https://bugzilla.mozilla.org/show_bug.cgi?id=969757. However, there are additions to it like this one
* https://bugzilla.mozilla.org/show_bug.cgi?id=974486 and this https://bugzilla.mozilla.org/show_bug.cgi?id=974370. This makes me think that there may be other removals and I really wonder how to best make sure they all are kept — checking each and every modified file is obviously a too much for a human being.

Perhaps I need to create a single giant diff between ESR24 and ESR31 and search for anything that look like `os2` or `OS/2` there.

**Comment by dryeo** [comment]
On 04/08/15 12:04 PM, Dmitriy Kuminov wrote:
> Perhaps I need to create a single giant diff between ESR24 and ESR31 and
> search for anything that look like |os2| or |OS/2| there.

Be better to apply the 1.4MB diff from 969757 to 31 and then do a diff 
between 29 and 31 otherwise it'll be a huge undertaking.

**Comment by dmik** [comment]
Yes, but this still leaves us a chance that we skip something. The diff between `esr24` and `esr31` is huge indeed (290MB) but I think that a simple search for `os2` and `os/2` there should work. I will also be looking into the original 1.4 M diff in parallel.

**Comment by dmik** [comment]
Important note: there will be a single commit that brings the ESR31 changes *and* restores all the OS/2 things they have cut out around version 29. This will also include numerous adaptions of the existing OS/2 code to the significantly changed parts (e.g. they have reorganized a lot in `toolkit/components/osfile`). The reason for it to be a single commit rather a series of the separate ones is because of many conflicting changes. E.g. there are many files that were removed upstream but have been changed by us during our OS/2 work. Applying the original ERS31 commit would mean deleting these files and then resurrecting them again somehow including all our previous patches. This looks much more complex than doing a single "all-inclusive" commit. After all, we will be always able to compare the original ESR31 diff with what we merge into our repo and see what has been changed in between.

The above doesn't mean though that the result will immediately build and run. It is impossible to take all the details into account w/o even compiling the code. But it will be a minor work I hope.

This also explains why it is taking so long. But see above.

**Comment by dmik** [comment]
The main part of merge is done — I've resolved all conflicts and restored many removals. Now the last part — checking through the 1.4 M diff and others to pick up the remaining OS/2 removals and get rid of  them in the final commit.

**Comment by dmik** [comment]
So, this historical merge is finally done, **49,288** changed files  with 13,717 additions and 8,144 deletions, see 8b9ba001473bcabdff75269287409968abb56bac. I tried to do as much OS/2 work as possible (including adapting new code to OS/2 at once as described above) but of course some glitches are still possible. This is about to be sorted during the build stage.

Before building, I have to go through my todo list and check things that I marked as requiring attention (new config variables, some questionable changes etc.).

**Comment by dmik** [comment]
The above relates to Mozilla bug 969757. The OS/2 code removed in bugs 974486 and 974379 was restored separately, in 23d9cd12cb0dabb604f76566c336d27b5762baf0.

**Comment by dmik** [comment]
And here we have the first problem with configure. Of course, it's virtualenv:
{{{
Installing setuptools, pip...
  Complete output from command D:/Coding/mozilla/ma...nv/bin/python2.7.exe -c "import sys, pip; sys...d\"] + sys.argv[1:]))" setuptools pip:
  Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "D:/Coding/mozilla/master/python/virtualenv/virtualenv_support/pip-1.5.4-py2.py3-none-any.whl/pip/__init__.py", line 9, in <module>
  File "D:/Coding/mozilla/master/python/virtualenv/virtualenv_support/pip-1.5.4-py2.py3-none-any.whl/pip/log.py", line 9, in <module>
  File "D:/Coding/mozilla/master/python/virtualenv/virtualenv_support/pip-1.5.4-py2.py3-none-any.whl/pip/_vendor/colorama/__init__.py", line 2, in <module>
  File "D:/Coding/mozilla/master/python/virtualenv/virtualenv_support/pip-1.5.4-py2.py3-none-any.whl/pip/_vendor/colorama/initialise.py", line 5, in <module>
  File "D:/Coding/mozilla/master/python/virtualenv/virtualenv_support/pip-1.5.4-py2.py3-none-any.whl/pip/_vendor/colorama/ansitowin32.py", line 6, in <module>
  File "D:/Coding/mozilla/master/python/virtualenv/virtualenv_support/pip-1.5.4-py2.py3-none-any.whl/pip/_vendor/colorama/winterm.py", line 2, in <module>
  File "D:/Coding/mozilla/master/python/virtualenv/virtualenv_support/pip-1.5.4-py2.py3-none-any.whl/pip/_vendor/colorama/win32.py", line 8, in <module>
  File "C:/USR/lib/python2.7/ctypes/__init__.py", line 453, in <module>
    pythonapi = PyDLL(None)
  File "C:/USR/lib/python2.7/ctypes/__init__.py", line 365, in __init__
    self._handle = _dlopen(self._name, mode)
NameError: global name '_dlopen' is not defined
----------------------------------------
...Installing setuptools, pip...done.
Traceback (most recent call last):
  File "D:/Coding/mozilla/master/python/virtualenv/virtualenv.py", line 2339, in <module>
    main()
  File "D:/Coding/mozilla/master/python/virtualenv/virtualenv.py", line 825, in main
    symlink=options.symlink)
  File "D:/Coding/mozilla/master/python/virtualenv/virtualenv.py", line 993, in create_environment
    install_wheel(to_install, py_executable, search_dirs)
  File "D:/Coding/mozilla/master/python/virtualenv/virtualenv.py", line 961, in install_wheel
    'PIP_NO_INDEX': '1'
  File "D:/Coding/mozilla/master/python/virtualenv/virtualenv.py", line 903, in call_subprocess
    % (cmd_desc, proc.returncode))
OSError: Command D:/Coding/mozilla/ma...nv/bin/python2.7.exe -c "import sys, pip; sys...d\"] + sys.argv[1:]))" setuptools pip failed with error code 1
Traceback (most recent call last):
  File "D:/Coding/mozilla/master/python/mozbuild/mozbuild/virtualenv.py", line 473, in <module>
    manager.ensure()
  File "D:/Coding/mozilla/master/python/mozbuild/mozbuild/virtualenv.py", line 128, in ensure
    return self.build()
  File "D:/Coding/mozilla/master/python/mozbuild/mozbuild/virtualenv.py", line 371, in build
    self.create()
  File "D:/Coding/mozilla/master/python/mozbuild/mozbuild/virtualenv.py", line 147, in create
    raise Exception('Error creating virtualenv.')
Exception: Error creating virtualenv.
}}}

**Comment by dmik** [comment]
They have updated the pip package from 1.2.1 to 1.5.4 and setuptools from 0.6.c11 to 2.2 in the meantime. So I think the problem is there. Digging it now.

**Comment by dmik** [comment]
The problem is a combination of factors. First, it's the new pip that uses the `colorama` library for console output. Second, this colorama library uses the `ctypes` standard python module that provides a bridge between python and C functions in native DLLs. This module apparently misses any OS/2 support. OTOH, I know that Yuri used the pyqt library that provides python interface to Qt and that works on OS/2... I need to collect more info.

The `colorama` module attempts to load the system Win32 DLLs to do powerful console control (colors etc) and fails to use the load function. If the function itself succeeded it would return `false` (no WIn32 DLL) and it should switch `colorama` to non-WIn32 mode IIGIC. I will try to add a stub that returns `false` to see if it works.

**Comment by dmik** [comment]
Ok, this dirty hack to `/usr/lib/python2.7/ctypes/__init__.py`:
``` diff
--- __init__.py.b   2015-02-26 21:23:20.000000000 +0400
+++ __init__.py 2015-04-23 21:55:04.000000000 +0400
@@ -140,6 +140,10 @@
 elif _os.name == "posix":
     from _ctypes import dlopen as _dlopen

+elif _os.name == "os2":
+    def _dlopen(name, mode):
+        return
+
 from _ctypes import sizeof, byref, addressof, alignment, resize
 from _ctypes import get_errno, set_errno
 from _ctypes import _SimpleCData
```
did the trick. I created a separate ticket for that task here: http://trac.netlabs.org/rpm/ticket/140.

Going further, `configure` now fails on the missing `libpulse` package. Investigating.

**Comment by dmik** [comment]
I got further through all configure stuff but it now fails upon the final Makefile generation step when processing `master/media/libyuv/moz.build`. The failure is this one:
```
The error appears to be part of the mozbuild.frontend.reader Python module itself! It is possible you have stumbled across a legitimate bug.

Traceback (most recent call last):
  File "D:/Coding/mozilla/master/python/mozbuild/mozbuild/frontend/reader.py", line 722, in read_mozbuild
    descend=descend, metadata=metadata):
  File "D:/Coding/mozilla/master/python/mozbuild/mozbuild/frontend/reader.py", line 819, in _read_mozbuild
    non_unified_sources = non_unified_sources):
  File "D:/Coding/mozilla/master/python/mozbuild/mozbuild/frontend/gyp_reader.py", line 111, in read_from_gyp
    params=params)
  File "D:/Coding/mozilla/master/media/webrtc/trunk/tools/gyp/pylib/gyp/__init__.py", line 132, in Load
    params['parallel'])
  File "D:/Coding/mozilla/master/media/webrtc/trunk/tools/gyp/pylib/gyp/input.py", line 2560, in Load
    variables, includes, depth, check, True)
  File "D:/Coding/mozilla/master/media/webrtc/trunk/tools/gyp/pylib/gyp/input.py", line 391, in LoadTargetBuildFile
    build_file_data, PHASE_EARLY, variables, build_file_path)
  File "D:/Coding/mozilla/master/media/webrtc/trunk/tools/gyp/pylib/gyp/input.py", line 1142, in ProcessVariablesAndConditionsInDict
    variables, build_file, 'variables')
  File "D:/Coding/mozilla/master/media/webrtc/trunk/tools/gyp/pylib/gyp/input.py", line 1142, in ProcessVariablesAndConditionsInDict
    variables, build_file, 'variables')
  File "D:/Coding/mozilla/master/media/webrtc/trunk/tools/gyp/pylib/gyp/input.py", line 1142, in ProcessVariablesAndConditionsInDict
    variables, build_file, 'variables')
  File "D:/Coding/mozilla/master/media/webrtc/trunk/tools/gyp/pylib/gyp/input.py", line 1194, in ProcessVariablesAndConditionsInDict
    ProcessConditionsInDict(the_dict, phase, variables, build_file)
  File "D:/Coding/mozilla/master/media/webrtc/trunk/tools/gyp/pylib/gyp/input.py", line 1071, in ProcessConditionsInDict
    variables, build_file)
  File "D:/Coding/mozilla/master/media/webrtc/trunk/tools/gyp/pylib/gyp/input.py", line 1149, in ProcessVariablesAndConditionsInDict
    expanded = ExpandVariables(value, phase, variables, build_file)
  File "D:/Coding/mozilla/master/media/webrtc/trunk/tools/gyp/pylib/gyp/input.py", line 872, in ExpandVariables
    cwd=build_file_dir)
  File "C:/USR/lib/python2.7/subprocess.py", line 709, in __init__
    errread, errwrite)
  File "C:/USR/lib/python2.7/subprocess.py", line 1326, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory
```
Looks like some OS/2 bits are missing again.

**Comment by dmik** [comment]
It turns out that python's `subprocess.Popen()` is broken on OS/2. At least here it fails to execute a complex command that involves piping (e.g. `uname | sed BLAH` like in the above case). Digging into python.

**Comment by dmik** [comment]
Created a python ticket for the above: http://trac.netlabs.org/rpm/ticket/141.

**Comment by dmik** [comment]
Okay, with the temporary workaround for Python (by replacing `/bin/sh` with `/@unixroot/usr/bin/sh` in `subprocess.py`) and a bunch of fixes in fd2ff6b0fa5c2838e32ad0f5c4ee823efbd63cb2 the configure stage now completes successfully.

I will try to build it now.

**Comment by dmik** [comment]
The build goes on. Fixing merge leftovers.

What really annoys me is that I get megabytes of this garbage now:
```
D:\Temp/ccerZFqd.s: Assembler messages:
D:\Temp/ccerZFqd.s:836: Warning: .stabs: description field '128bd' too big, try a different debug format
```

This comes from `gas` and seems to be somehow related to the compiler switch (from 4.7.3 to 4.9.2).

**Comment by dmik** [comment]
Passing `-Wa,-W` to `gcc` should shut up `gas` but the problem is that I don't know which file produces such an assembly (I need to build in single-thread mode to know this). Will leave it for later.

**Comment by dmik** [comment]
Faced the first more or less problem: libvpx was updated and requires more OS/2 porting. I did it mostly, but now fight with `nasm` that doesn't like one new `.asm` file:
```
nasm -o vp9_copy_sse2.o -f obj -I. -ID:/Coding/mozilla/master/media/libvpx/ -ID:/Coding/mozilla/master/media/libvpx/vpx_ports/  -g    D:/Coding/mozilla/master/m
edia/libvpx/vp9/common/x86/vp9_copy_sse2.asm
D:/Coding/mozilla/master/media/libvpx/vp9/common/x86/vp9_copy_sse2.asm:151: error: symbol `eaxm' undefined
D:/Coding/mozilla/master/media/libvpx/vp9/common/x86/vp9_copy_sse2.asm:151: error: symbol `ecxm' undefined
D:/Coding/mozilla/master/media/libvpx/vp9/common/x86/vp9_copy_sse2.asm:151: error: symbol `edxm' undefined
D:/Coding/mozilla/master/media/libvpx/vp9/common/x86/vp9_copy_sse2.asm:151: error: symbol `ebxm' undefined
D:/Coding/mozilla/master/media/libvpx/vp9/common/x86/vp9_copy_sse2.asm:152: error: symbol `eaxm' undefined
D:/Coding/mozilla/master/media/libvpx/vp9/common/x86/vp9_copy_sse2.asm:152: error: symbol `ecxm' undefined
D:/Coding/mozilla/master/media/libvpx/vp9/common/x86/vp9_copy_sse2.asm:152: error: symbol `edxm' undefined
D:/Coding/mozilla/master/media/libvpx/vp9/common/x86/vp9_copy_sse2.asm:152: error: symbol `ebxm' undefined
```

**Comment by dryeo** [comment]
On 04/29/15 05:34 PM, Dmitriy Kuminov wrote:
> but now fight with |nasm| that doesn't like one new |.asm| file:

The NASM included with RPM/YUM is old, try upgrading. There's a newer 
one on Hobbes or the latest which I believe was forgotten to be released 
but available in their git repository.
Officially libvpx does (did?) support both NASM and YASM

**Comment by dryeo** [comment]
There's also KO Myung-Hun port of libvpx, http://hobbes.nmsu.edu/download/pub/os2/dev/mm/libvpx-v1.3.0.zip. Probably pretty complete and he is good at adapting YASM code to NASM.

**Comment by dryeo** [comment]
Perhaps this,
```
diff --git a/media/libvpx/third_party/x86inc/x86inc.asm b/media/libvpx/third_party/x86inc/x86inc.asm
index 9e4918a..afba317 100644
--- a/media/libvpx/third_party/x86inc/x86inc.asm
+++ b/media/libvpx/third_party/x86inc/x86inc.asm
@@ -234,10 +234,10 @@ ALIGNMODE k7
         %define r%1mp %2
     %elif ARCH_X86_64 ; memory
         %define r%1m [rsp + stack_offset + %6]
-        %define r%1mp qword r %+ %1m
+        %define r%1mp qword r %+ %1 %+ m
     %else
         %define r%1m [esp + stack_offset + %6]
-        %define r%1mp dword r %+ %1m
+        %define r%1mp dword r %+ %1 %+ m
     %endif
     %define r%1  %2
 %endmacro
```

**Comment by dmik** [comment]
Yes, I know that our nasm is old, but thanks for pointing out, Dave! FF may be built using the system libvpx and that's what we will certainly do at some point.

**Comment by dmik** [comment]
BTW, both the KOMH port and FF 31 libvpx are 1.3.0 which means we can take KOMH patches for now.

**Comment by dmik** [comment]
Actually there is not much to apply from the KOMH work, only the above .asm fix, the rest is either not related or arguable. Anyway, with 1f951f59a66081f1ca0ff3e443357098caeab38c and 9651c584c2805a2360a853a04156b567b91218e8 `libvpx` now builds.

**Comment by dmik** [comment]
A short report: going further step by step (refer to the commits). The current task is adopting the widget code and in particular, the `nsWindow` class. There are lots of changes over there.

**Comment by dmik** [comment]
BTW, widget code seems to be done, there is really not much left (I hope).

**Comment by dmik** [comment]
Note that in 7a04051ed7a238a9259d489dc3bf66f94464abea I disabled the Profiler code (accidentally enabled in 9732dbb1f4c5085fad34b9a4d068119ded894b3c) again as it is not ported to OS/2. I doubt that we will need it soon so I postpone this for later in #103.

**Comment by dmik** [comment]
It seems that everything is successfully built by now. There are some failures at link stage (due to missing symbols) but that's expected. Will fix them next.

**Comment by dmik** [comment]
Attempting to link `XUL.DLL` showed that there is a number new classes and methods that need to be implemented (e.g. `VolatileBuffer`, `GetAppliactionParametersFrom(ACString)`). Shouldn't be a big deal.

**Comment by dmik** [comment]
I have added all missing stuff and there is no unresolved externals any more, however the link stage still fails with Error 1 and no message. This smells like a `wlink` problem... I already saw it back then when it was not capable of handling too many objects. I need to investigate more.

**Comment by dmik** [comment]
Well, the wlink error is in the .map file, actually:
```
Error! E2021: size of segment text exceeds 64k by 21899 bytes (which is 0x558b).
```
Judging from the rest of map file, it refers to the following AUTO segment:
```
text                                  AUTO           0003:00000000   0001558b
rodata                                AUTO           0004:00000000   00000470
```
I have no idea why this segment is created now; it is not present in the previous build of Mozilla (which is the same GGG 4.9.2 compiler so I doubt the compiler is in vain). I need to do some research.

**Comment by StevenLevine** [comment]
I suspect something (probably the compiler) is generating 16-bit segments that should be 32-bit segments.

**Comment by dryeo** [comment]
On 05/22/15 01:26 PM, StevenLevine wrote:
> I suspect something (probably the compiler) is generating 16-bit 
> segments that should be 32-bit segments.

Actually as usual it is the assembler, we didn't finish porting to obj format. With this patch I successfully built xul.dll. Wish github supported attachments
```
From 94e7d872edf400f7857025d9a45649a2a2b62991 Mon Sep 17 00:00:00 2001
From: Dave Yeo <dave.r.yeo@gmail.com>
Date: Sat, 23 May 2015 16:46:56 -0700
Subject: [PATCH] Fix libvpx to use USE32 and FLAT to avoid 16bit segments

---
 media/libvpx/third_party/x86inc/x86inc.asm          | 3 +++
 media/libvpx/vp9/common/x86/vp9_copy_sse2.asm       | 4 ++++
 media/libvpx/vp9/common/x86/vp9_intrapred_sse2.asm  | 5 +++++
 media/libvpx/vp9/common/x86/vp9_intrapred_ssse3.asm | 4 ++++
 4 files changed, 16 insertions(+)

diff --git a/media/libvpx/third_party/x86inc/x86inc.asm b/media/libvpx/third_party/x86inc/x86inc.asm
index afba317..e1905e2 100644
--- a/media/libvpx/third_party/x86inc/x86inc.asm
+++ b/media/libvpx/third_party/x86inc/x86inc.asm
@@ -83,6 +83,9 @@
         fakegot:
     %elifidn __OUTPUT_FORMAT__,aout
         section .text
+    %elifidn __OUTPUT_FORMAT__,obj
+       section .text  align=16 public use32 class=CODE
+       section  .rodata align=16 public use32 class=DATA
     %else
         SECTION .rodata align=%1
     %endif
diff --git a/media/libvpx/vp9/common/x86/vp9_copy_sse2.asm b/media/libvpx/vp9/common/x86/vp9_copy_sse2.asm
index dd522c6..d9d8987 100644
--- a/media/libvpx/vp9/common/x86/vp9_copy_sse2.asm
+++ b/media/libvpx/vp9/common/x86/vp9_copy_sse2.asm
@@ -10,7 +10,11 @@
 
 %include "third_party/x86inc/x86inc.asm"
 
+%ifidn __OUTPUT_FORMAT__,obj
+SECTION .text  align=16 public use32 class=CODE
+%else
 SECTION .text
+%endif
 
 %macro convolve_fn 1
 INIT_XMM sse2
diff --git a/media/libvpx/vp9/common/x86/vp9_intrapred_sse2.asm b/media/libvpx/vp9/common/x86/vp9_intrapred_sse2.asm
index 69b07f6..7c3f529 100644
--- a/media/libvpx/vp9/common/x86/vp9_intrapred_sse2.asm
+++ b/media/libvpx/vp9/common/x86/vp9_intrapred_sse2.asm
@@ -16,7 +16,12 @@ pw_8:  times 8 dw 8
 pw_16: times 8 dw 16
 pw_32: times 8 dw 32
 
+%ifidn __OUTPUT_FORMAT__,obj
+SECTION .text  align=16 public use32 class=CODE
+%else
 SECTION .text
+%endif
+
 
 INIT_MMX sse
 cglobal dc_predictor_4x4, 4, 5, 2, dst, stride, above, left, goffset
diff --git a/media/libvpx/vp9/common/x86/vp9_intrapred_ssse3.asm b/media/libvpx/vp9/common/x86/vp9_intrapred_ssse3.asm
index 88df9b2..e69ec87 100644
--- a/media/libvpx/vp9/common/x86/vp9_intrapred_ssse3.asm
+++ b/media/libvpx/vp9/common/x86/vp9_intrapred_ssse3.asm
@@ -31,7 +31,11 @@ sh_bfedcba9876543210: db 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0
 sh_b1233: db 1, 2, 3, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
 sh_b2333: db 2, 3, 3, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
 
+%ifidn __OUTPUT_FORMAT__,obj
+SECTION .text  align=16 public use32 class=CODE
+%else
 SECTION .text
+%endif
 
 INIT_MMX ssse3
 cglobal h_predictor_4x4, 2, 4, 3, dst, stride, line, left
-- 
2.0.0
```

Hopefully formatting is correct, github not loading so using email.

**Comment by dryeo** [comment]
The line section .rodata... is not needed in x86inc.asm. Seems .text is 
not ideal but changing it to .rodata or .data causes linking xul.dll to 
fail due to 16bit segment

**Comment by dryeo** [comment]
Next build break is building brwsrcmp.dll.
Need to do s/bin/lib/ in configure.in for XPCOM_FROZEN_LDOPTS and MOZ_COMPONENT_NSPR_LIBS
Re-add SHORT_LIBNAME = brwsrcmp to browser/components/build/Makefile.in and remove SHORT_LIBNAME from the list of depreciated variables in config.mk
Hopefully speed things up as it takes forever to build here.

**Comment by dmik** [comment]
Dave, thanks for the hint, I guessed something like that. I committed a complete (I hope) solution in c99319a1b8f0f185297c61ffa4124fd8ded1c58c. With it, we don't have any extra `text` or `rodata` segments anymore, everything lives in the standard `TEXT32` segment now (as for the normal code produced by C/C++). Going further. Will take your comments on `brwsrcmp` into account.

**Comment by dmik** [comment]
Finally, I have everything built. It doesn't start though yet, crashes with this in popuplog right before showing the window:
```
05-26-2015  11:17:14  SYS3175  PID fe87  TID 0001  Slot 00a9
D:\CODING\MOZILLA\MASTER-BUILD\DIST\BIN\FIREFOX.EXE
c0000005
14a13c7a
P1=00000001  P2=00000010  P3=XXXXXXXX  P4=XXXXXXXX
EAX=00000000  EBX=00000000  ECX=1bc00002  EDX=2003013c
ESI=174126bc  EDI=00000000
DS=0053  DSACC=d0f3  DSLIM=9fffffff
ES=0053  ESACC=d0f3  ESLIM=9fffffff
FS=150b  FSACC=00f3  FSLIM=00000030
GS=0000  GSACC=****  GSLIM=********
CS:EIP=005b:1d84e554  CSACC=d0df  CSLIM=9fffffff
SS:ESP=0053:0012fd6c  SSACC=d0f3  SSLIM=9fffffff
EBP=0012fdb8  FLG=00010202

XUL.DLL 0001:00a63c7a
```
But that's kinda expected. I think it's time to cook a debug build to get some logging.

**Comment by dmik** [comment]
Actually, w/o the debug build I already get something interesting. First, this is in stderr:
```
1432642878795	addons.xpi-utils	WARN	Extensions database D:\Users\mozdev\Mozilla\Firefox\Profiles\pycia5on.default\extensions.json exists but is not readable; rebuilding: [Exception... "Component returned failure code: 0x80004005 (NS_ERROR_FAILURE) [nsIFileInputStream.init]"  nsresult: "0x80004005 (NS_ERROR_FAILURE)"  location: "JS frame :: resource://gre/modules/addons/XPIProvider.jsm -> resource://gre/modules/addons/XPIProviderUtils.js :: XPIDB_syncLoadDB :: line 540"  data: no] Stack trace: XPIDB_syncLoadDB()@resource://gre/modules/addons/XPIProvider.jsm -> resource://gre/modules/addons/XPIProviderUtils.js:540 < XPI_checkForChanges()@resource://gre/modules/addons/XPIProvider.jsm:3423 < XPI_startup()@resource://gre/modules/addons/XPIProvider.jsm:1884 < AMI_callProviders()@resource://gre/modules/AddonManager.jsm:869 < AMI_startup()@resource://gre/modules/AddonManager.jsm:745 < AMP_startup()@resource://gre/modules/AddonManager.jsm:2318 < AMC_observe()@addonManager.js:55 < <file:unknown>
1432642882043	addons.xpi	ERROR	Failed to process extension changes at startup: ReferenceError: OS is not defined (resource://gre/modules/osfile/osfile_unix_allthreads.jsm:44) JS Stack trace: @osfile_unix_allthreads.jsm:44:1 < @osfile_async_front.jsm:44:3 < @osfile.jsm:11:3 < @DeferredSave.jsm:11:1 < XPCU_moduleLambda@XPCOMUtils.jsm:226:7 < XPCU_defineLazyGetter/<.get@XPCOMUtils.jsm:177:9 < this.XPIDatabase.saveChanges@XPIProviderUtils.js:432:7 < XPIDB_makeAddonVisible@XPIProviderUtils.js:1329:5 < XPIDB_addAddonMetadata@XPIProviderUtils.js:1269:7 < addMetadata@XPIProvider.jsm:3076:11 < XPI_processFileChanges@XPIProvider.jsm:3255:1 < XPI_checkForChanges@XPIProvider.jsm:3428:1 < XPI_startup@XPIProvider.jsm:1884:1 < AMI_callProviders@AddonManager.jsm:869:11 < AMI_startup@AddonManager.jsm:745:1 < AMP_startup@AddonManager.jsm:2318:5 < AMC_observe@addonManager.js:55:7
DIVE is disabled - Panorama's shadow-buffer is enabled
*************************
A coding exception was thrown and uncaught in a Task.

Full message: ReferenceError: OS is not defined
Full stack: @resource://gre/modules/osfile/osfile_unix_allthreads.jsm:44:1
@resource://gre/modules/osfile/osfile_async_front.jsm:44:3
@resource://gre/modules/osfile.jsm:11:3
@resource://gre/modules/PlacesBackups.jsm:17:1
XPCU_moduleLambda@resource://gre/modules/XPCOMUtils.jsm:226:7
XPCU_defineLazyGetter/<.get@resource://gre/modules/XPCOMUtils.jsm:177:9
BG__initPlaces/<@file:///D:/CODING/MOZILLA/MASTER-BUILD/DIST/BIN/browser/components/nsBrowserGlue.js:1170:11
TaskImpl_run@resource://gre/modules/Task.jsm:298:13
TaskImpl@resource://gre/modules/Task.jsm:247:3
createAsyncFunction/asyncFunction@resource://gre/modules/Task.jsm:224:7
Task_spawn@resource://gre/modules/Task.jsm:139:5
BG__initPlaces@file:///D:/CODING/MOZILLA/MASTER-BUILD/DIST/BIN/browser/components/nsBrowserGlue.js:1200:1
BG_observe@file:///D:/CODING/MOZILLA/MASTER-BUILD/DIST/BIN/browser/components/nsBrowserGlue.js:237:11

*************************
Creating 12E7_01.TRP
```
And this in TRP:
```
 Filename: D:\CODING\MOZILLA\MASTER-BUILD\DIST\BIN\XUL.DLL 05/26/2015 01:52:18 59,083,908
 Address:  005B:10973C7A (0001:00A63C7A)
 Cause:    Attempted to read from 00000010
           (not a valid address)
...
 --------  ---------  --------  -------------  -----------------------
 Trap  ->  10973C7A   XUL       0001:00A63C7A  Matrix.h#269 __ZN7mozilla6layers20RotatedContentBuffer15IsClippingCheapEPNS_3gfx10DrawTargetERK11nsIntRegion + A 0001:00A63C70 (D:\Coding\mozilla\master\gfx\layers\RotatedBuffer.cpp)

```

**Comment by dmik** [comment]
Ok, the `OS` exception has gone now (see cd1efee183fea88a2473e2466735d758a407358b), looking for the rest. As for the trap, it seems that `RotatedContentBuffer::IsClippingCheap(DrawTarget* aTarget, const nsIntRegion& aRegion)` in `RotateBuffer.cpp` is called with `aTarget` being `NULL` for an unknown reason (yet).

**Comment by dryeo** [comment]
First run I get
  Trap  ->  13BF4D4A   XUL       0001:009F4D4A  between 
mozilla::layers::RotatedContentBuffer::IsClippingCheap + A and 
mozilla::layers::RotatedContentBuffer::DrawTo - 116  (both in 
RotatedBuffer.o)

Which is probably the same as you.

**Comment by dmik** [comment]
Yes, I bet it's the same. Strange that you have different numbers. What build type are you using?

**Comment by dryeo** [comment]
On 05/26/15 08:50 AM, Dmitriy Kuminov wrote:
> Yes, I bet it's the same. Strange that you have different numbers. What
> build type are you using?

Not sure what you mean by build type. I'm building FF and the different 
numbers may be due to targeting i686 rather then i486.
Never did have your first error and I note that the choose profile 
window is mostly drawn when the crash happens (I have FF & SM set to 
permanently start with the Profile Manager)

**Comment by dmik** [comment]
This is what I meant. Ok. Regarding the crash. For some reason, a `gfxContext` instance passed to `BasicLayerManager::PaintSelfOrChildren` is a cairo-based one, but the code thinks it's a Thebes based one and crashes because the Thebes target is not set. I need to deeply learn the program logic to find why.

**Comment by dryeo** [comment]
Another problem is that <-mstackrealign> seems not to be being used when <-sse2> is used

**Comment by dmik** [comment]
Yes, the -mstackrealign thingy is in my internal todo list. I didn't do it on merge so that it can be checked once again and committed in a separate step when everything builds.

**Comment by dmik** [comment]
And here we are! The first run of Firefox 31.6.0 on OS/2!

![2015-06-02 15 33 50](https://cloud.githubusercontent.com/assets/900726/7935846/d3607e3a-093c-11e5-95dd-b236dee878fe.png)

The commits will follow.

**Comment by dbanet** [comment]
Congratulations, Dmitry!

I was really rather afraid that porting Australis-based Firefox will take much-much more time and effort.

**Comment by dmik** [comment]
I doubt it has a lot to do with the Austrails theme. It's more about some redesign within the gfx/thebes subsystem. The fix for this particular crash turned out to be rather tiny: dc9b1ee812ee975b1dba1df3b63f85b8d45b2f7b. But it's necessary to mention that I already did some adaptation to the new gfx code when merging esr31.

The thing is still not very stable yet. Things to improve in the first place:
- Use proper stack align for SSE (I see traps in SSE code).
- Fix JS erros we see at startup:
```
1433258519533	addons.xpi-utils	WARN	Extensions database D:\Users\mozdev\Mozilla\Firefox\Profiles\adc8cjim.default\extensions.json exists but is not readable; rebuilding: [Exception... "Component returned failure code: 0x80004005 (NS_ERROR_FAILURE) [nsIFileInputStream.init]"  nsresult: "0x80004005 (NS_ERROR_FAILURE)"  location: "JS frame :: resource://gre/modules/addons/XPIProvider.jsm -> resource://gre/modules/addons/XPIProviderUtils.js :: XPIDB_syncLoadDB :: line 540"  data: no] Stack trace: XPIDB_syncLoadDB()@resource://gre/modules/addons/XPIProvider.jsm -> resource://gre/modules/addons/XPIProviderUtils.js:540 < XPI_checkForChanges()@resource://gre/modules/addons/XPIProvider.jsm:3423 < XPI_startup()@resource://gre/modules/addons/XPIProvider.jsm:1884 < AMI_callProviders()@resource://gre/modules/AddonManager.jsm:869 < AMI_startup()@resource://gre/modules/AddonManager.jsm:745 < AMP_startup()@resource://gre/modules/AddonManager.jsm:2318 < AMC_observe()@addonManager.js:55 < <file:unknown>
DIVE is disabled - Panorama's shadow-buffer is enabled
console.error: 
  Message: Unix error 110 during operation open on file D:\Users\mozdev\Mozilla\Firefox\Profiles\adc8cjim.default\sessionstore.js (Unknown error 110)
```
- Fix JS errors we see at termination:
```
WARNING: A phase completion condition is taking too long to complete. Condition: OS.File: flush I/O queued before profile-before-change Phase: profile-before-change State: (none)
WARNING: A phase completion condition is taking too long to complete. Condition: AddonManager: shutting down providers Phase: profile-before-change State: (none)
WARNING: A phase completion condition is taking too long to complete. Condition: CrashMonitor: Writing notifications to file after receiving profile-before-change Phase: profile-before-change State: (none)
ERROR: At least one completion condition failed to complete within a reasonable amount of time. Causing a crash to ensure that we do not leave the user with an unresponsive process draining resources. Conditions: [{"name":"OS.File: flush I/O queued before profile-before-change","state":{"launched":true,"shutdown":false,"worker":true,"pendingReset":false,"latestSent":["Tue Jun 02 2015 19:23:45 GMT+0400 (MSK)","makeDir",[{"string":"D:\\Users\\mozdev\\Mozilla\\Firefox\\Profiles\\adc8cjim.default\\bookmarkbackups"},{"ignoreExisting":true}]],"latestReceived":null,"messagesSent":0,"messagesReceived":1,"messagesQueued":9,"DEBUG":false}},{"name":"AddonManager: shutting down providers","state":"(none)"},{"name":"CrashMonitor: Writing notifications to file after receiving profile-before-change","state":"(none)"}] Phase: profile-before-change
WARNING: No crash reporter available
[72] ###!!! ABORT: file resource://gre/modules/AsyncShutdown.jsm, line 431
[72] ###!!! ABORT: file resource://gre/modules/AsyncShutdown.jsm, line 431
```
- Fix about:home URL error:
```
XML Parsing Error: error in processing external entity reference Location: file:///D:/CODING/MOZILLA/MASTER-BUILD/DIST/BIN/browser/chrome/browser/content/browser/abouthome/aboutHome.xhtml Line Number 11, Column 3: %htmlDTD; --^
```
- Sort out JIT crashes: https://gist.github.com/dmik/0eb4dd3b31eaf3887aac.

**Comment by wztest** [comment]
Great to see ESR 31 running on OS/2 :-) Hopefully not too much bugs left till the beta build.

**Comment by dryeo** [comment]
Using my usual FF profile, I got a different JS crash before the window was even drawn. Using a less used profile gave me most of the problems that you list above and exiting a crash in
 Trap  ->  1CE6047E   MOZALLOC  0001:0000047E  between mozalloc_abort + 2E and mozalloc_handle_oom - 32  (in mozalloc_abort.o and mozalloc_oom.o)
with the top label
 0012C1B4  138C0F28   XUL       0001:00040F28  between nsDebugImpl::SetMultiprocessMode + 28 and _NS_DebugBreak - 148  (both in nsDebugImpl.o)

They've sure sped up their development compared to earlier. I'll build SM now and see if some of these problems are actually FF or as more likely xul.

**Comment by dmik** [comment]
The `-mstackrealign` option is back in 88e1b35a579fb699cd7f43da7e3015d4065e640d. Should be no more crashes in SSE2 code. Note that I added `-mstackrealign` for `-mssse3` and `-mavx` as well as these also deal with 128 bit registers.

**Comment by dmik** [comment]
The crash in JS code seems to be the crash in the generated JIT code. Frequently happens in the same place. I will work on that first since it's the major issue so far.

Dave, your crash is from MOZ_CRASH() which is intentionally done by the guarding code because some shutdown task is taking too long to complete.

**Comment by dmik** [comment]
From what I learned so far, the JS crashes in the new Ion code. Ion is the JIT implementation for JS used in Mozilla. This is a brand new part that didn't exist in FF24 and before. My findings show that the generated code sometimes produces weird address and offset values that cause unexpected access violation when accessing memory. I need to debug this little bit more.

I successfully built the debug build of Firefox (to get more debug information and also to have real functions instead of inlined ones for better call stack tracing). Now I need to fix the debug build because it terminates on assertions in several places (meaning something is going wrong).

**Comment by dmik** [comment]
I got the debug build up and running now but unfortunately this doesn't give me any new information regarding the Ion crash.

**Comment by dmik** [comment]
Actually, I seem to have the crash always at the same place with the debug build.This is already something. The code is rather complex though, I don't have much experience with JIT.

**Comment by dmik** [comment]
I found what was wrong. The new Ion (JIT) engine implements the so-called interrupt technique to kick into the JIT code being executed when it needs to execute some hook. On Win this is done through setting JIT code's memory to "no access" mode which causes the access violation exception next time this code is executed. The exception handler catches this, restores memory access rights to "all access", executes necessary hooks, and lets the JIT code continue from the place it was interrupted at.

As on OS/2 it's impossible to set "no access" mode for memory pages other than decommit them, this is what I did when I discovered this new code when merging esr31 into trunk — w/o knowing back then on how that code was supposed to be used. Apparently, this didn't work as designed because decimating pages releases storage and when they are committed again they are filled with zeroes (thus, effectively killing all JIT code that was there before). Now that I realized this, I had to use a different technique on OS/2 that involves the PAG_GUARD page flag and the XCPT_GUARD_PAGE_VIOLATION exception (instead of XCPT_ACCESS_VIOLATION).

I gave it quite a long run (including gmail access) and no JS crashes any more. I conclude that this particular problem is solved. The fix is in 7137167cfd18e66abb31e36e5fb6e0350d49bce8. Moving to other problems now (mentioned above).

**Comment by dmik** [comment]
Note that in 92f2673146f82838e91443df560ce67dd6f36cff I also made all DosAlloc allocations obey --disable-os2-high-mem so that no OBJ_ANY is used in such a case (just while I'm at this code — as practice shows our IBM debuggers are unable to disassemble code from high memory and sometimes it may be desirable to disable high memory completely).

**Comment by komh** [comment]
configure fails due to python2.7. I should symlink python2.7.exe to python2.7. Is this required task ?

And build is performed like the first time whenever running make. So after fixing some problems, I had to wait the same time, about 3~4 hours, it took during previous building. 

Is this right ? Or I'm doing something wrong ?

**Comment by komh** [comment]
And I got when configuring.

```
Installing setuptools, pip...done.

Killed by SIGSEGV
pid=0x033c ppid=0x033b tid=0x0001 slot=0x00ae pri=0x0200 mc=0x0001 ps=0x0010
G:\LANG\WORK\MOZILLA\MOZILLA-OS2.GIT\OBJ-FF\_VIRTUALENV\BIN\PYTHON2.7
PTHR01 0:000015c4
cs:eip=005b:1d0315c4      ss:esp=0053:0012ed50      ebp=0012ed68
 ds=0053      es=0053      fs=150b      gs=0000     efl=00012206
eax=1c87b444 ebx=1bff0030 ecx=00140002 edx=00000001 edi=20202b30 esi=00000000
Process dumping was disabled, use DUMPPROC / PROCDUMP to enable it.

Error processing command. Ignoring because optional. (optional:setup.py:python/p
sutil:build_ext:--inplace)
```
Is this OK ?

As well as,  why pthread_xxx messages are printed ?

```
pthread_mutex_lock[1]: Error 0
pthread_mutex_lock[3]: Error 0
pthread_cond_destroy: Error 0
pthread_mutex_lock[1]: Error 0
pthread_mutex_lock[3]: Error 0
pthread_cond_destroy: Error 0
pthread_mutex_lock[1]: Error 0
pthread_mutex_lock[3]: Error 0
pthread_cond_destroy: Error 0
pthread_mutex_lock[1]: Error 0
pthread_mutex_lock[3]: Error 0
pthread_cond_destroy: Error 0
pthread_mutex_lock[1]: Error 0
pthread_mutex_lock[3]: Error 0
pthread_cond_destroy: Error 0
pthread_mutex_lock[1]: Error 0
pthread_mutex_lock[3]: Error 0
pthread_cond_destroy: Error 0
pthread_mutex_lock[1]: Error 0
pthread_mutex_lock[3]: Error 0
pthread_cond_destroy: Error 0
pthread_mutex_lock[1]: Error 0
pthread_mutex_lock[3]: Error 0
pthread_cond_destroy: Error 0
```

Finally, I cannot redirect the output of configure like **configure 2>&1 | tee configure.log**.  If redirecting, configure fails with some weird errors.

**Comment by dmik** [comment]
Yes, AFAIR python2.7 is a requirement. Please keep in mind that configure expects a decent unix environment and that I optimize its OS/2 parts to work under the RPM/YUM environment.

And yes, they changed something in the root make file so that if you start `make` from root w/o any arguments, you will always get a complete rebuild (and that sucks big time). Maybe there is some target that lets bypass that but I haven't figured out it yet. When I debug/modify stuff, I build from individual subdirs where I change files (with `make` or `make libs`) and, since in most cases this will generate a static library, I go to `toolkit/library` to rebuild XUL.DLL to pull the new library in.

**Comment by dmik** [comment]
Regarding your crash, it seems that you have to update your PTHR01.DLL (or python or something else). With the current RPM versions there are no crashes here. The pthread messages are some debug leftovers in python IIRC (Yuri promised to remove them in the next build).

Redirection works perfect here with the latest ASH from RPM, I use it all the time.

**Comment by komh** [comment]
I'm using latest RPM/YUM environment. However, python2.7 did not exist. Later, install python2.7 as well when installing python2.7 package.

Sucks mozilla build system! Wasted too much my time.

As I said above, I'm using latest RPM/YUM environment. **yum update** do all things. doesn't it ?

Redirection seems to be a problem of 4OS2(tee). Now use tee in shell script, and works well. 

Thanks.

**Comment by komh** [comment]
Ah, I forgot one. I failed to build xul.dll due to E3009 of wlink. At that time, I was using physical memory 2GB and setting VAL to 1536. 

What is the proper size for physical memory and VAL ? I saw 2560 for VAL in other issues. 

I think, it would be better to add this information to developer wiki.

**Comment by dmik** [comment]
I can't say for sure why you didn't have the right symlink for python2.7 but I'm glad that (re?)installing it solved the issue for you. I totally second you: the build system is a whole crap. But we have to leave with it as it turned out that converting it to something else is as complex as doing half of Firefox from scratch (I already tried converting to kmk).

Yes, `yum update` usually does all you need provided that you have proper repositories set up. There are some bugs for sure but for most it works out of the box. Feel free to ask me or Yuri if you have problems.

Yes, it's assumed that everything is run under `sh.exe` (as I wrote before).

And you are right, from my experience VAL should be at least 2560 in order for the link stage to succeed. But that may be platform-dependent. AFAIR for Dave 2048 works.

**Comment by dryeo** [comment]
Redirection works fine here as well running from cmd.exe. I think that 
mach is supposed to get around the building everything thing and is now 
the official way to build.
I didn't need to adjust VAL from the default until building a debug 
build where I jacked it to the 3072. This will vary from system to system.
As for your trap 8, did you try building with -j1? My system traps when 
building Thunderbird if I use higher then -j1 whereas SM and FF are 
fine. I also build SM and TB the old fashioned way with client.mk and 
%MOZCONFIG%.
BTW SeaMonkey is working here, needs work with a couple of scripts, 
probably in toolkit extensions and some osfile updates as complains 
about extensions.jsom and the tools menu doesn't even work. about:addons 
does. Also hangs on exit until it kills itself with an INT 3 in 
mozalloc, nice feature.

**Comment by dryeo** [comment]
Testing FF I see FF has most of the same issues as SM

**Comment by dmik** [comment]
Dave, yes, the mach frontend is new and I didn't try it yet. Let's hope it presents something better than raw `make` does. I confirm the issues you mention (see [#issuecomment-107995179](#issuecomment-107995179) above).

**Comment by StevenLevine** [comment]
>Ah, I forgot one. I failed to build xul.dll due to E3009 of wlink. At
>that time, I was using physical memory 2GB and setting VAL to 1536.

This should be OK for a retail build.  Just in case, make sure you are
using the most recent wlink build that reports itself as

** EXPERIMENTAL (HLL) ** Open Watcom Linker Version 2.0beta1 Limited
Availability

**Comment by komh** [comment]
I'm trying to build a release build, and using right that version of wl.

**Comment by komh** [comment]
This time, unfortunately, I got 

```
make[3]: Entering directory `G:/lang/work/mozilla/mozilla-os2.git/obj-ff/db/sqlite3/src'
rm -f mozsqlt3.def
echo LIBRARY mozsqlt3 INITINSTANCE TERMINSTANCE > mozsqlt3.def
echo PROTMODE >> mozsqlt3.def
echo CODE    LOADONCALL MOVEABLE DISCARDABLE >> mozsqlt3.def
echo DATA    PRELOAD MOVEABLE MULTIPLE NONSHARED >> mozsqlt3.def
echo EXPORTS >> mozsqlt3.def
G:/lang/work/mozilla/mozilla-os2.git/obj-ff/_virtualenv/bin/python -m mozbuild.action.preprocessor -DSQLITE_SECURE_DELETE=1 -DSQLITE_THREADSAFE=1 -DSQLITE_CORE=1 -DSQLITE_ENABLE_FTS3=1 -DSQLITE_ENABLE_UNLOCK_NOTIFY=1 -DSQLITE_DEFAULT_PAGE_SIZE=32768 -DSQLITE_MAX_DEFAULT_PAGE_SIZE=32768 -DSQLITE_MAX_SCHEMA_RETRY=25 -DMOZ_OS2_HIGH_MEMORY -DNO_NSPR_10_SUPPORT G:/lang/work/mozilla/mozilla-os2.git/db/sqlite3/src/sqlite.def | sed -e '1,/^EXPORTS$/ d' -e 's,sqlite3,_\0,' -e 's,\ DATA.*$,,' >> mozsqlt3.def
mozsqlt3.dll
rm -f mozsqlt3.dll
G:/lang/work/mozilla/mozilla-os2.git/obj-ff/_virtualenv/bin/python G:/lang/work/mozilla/mozilla-os2.git/config/expandlibs_exec.py --depend .deps/mozsqlt3.dll.pp --target mozsqlt3.dll --uselist --  gcc.exe -idirafter f:/lang/os2tk45/h -Wall -Wpointer-arith -Wdeclaration-after-statement -Werror=return-type -Werror=int-to-pointer-cast -Wtype-limits -Wempty-body -Wsign-compare -Wno-unused -Wcast-align -march=i486 -mtune=i686 -std=gnu99 -fgnu89-inline -fno-strict-aliasing -Zomf -fno-math-errno -pthread  -DNDEBUG -DTRIMMED -g -O3 -fomit-frame-pointer  -Zdll -Zhigh-mem -o mozsqlt3.dll  sqlite3.o   -lpthread  -Zomf -Zmap -Zlinker 'DISABLE 1121' -Zhigh-mem  -g -Zlinker 'option symfile=mozsqlt3.dbg'   -Zhigh-mem    -LG:/lang/work/mozilla/mozilla-os2.git/obj-ff/dist/lib -lmozglue -lmmap -Zomf  mozsqlt3.def  

Killed by SIGSEGV
pid=0x0a72 ppid=0x0a71 tid=0x0001 slot=0x00b5 pri=0x0200 mc=0x0001 ps=0x0010
G:\LANG\WORK\MOZILLA\MOZILLA-OS2.GIT\OBJ-FF\_VIRTUALENV\BIN\PYTHON2.7
PTHR01 0:000015c4
cs:eip=005b:1d0115c4      ss:esp=0053:0012f000      ebp=0012f018
 ds=0053      es=0053      fs=150b      gs=0000     efl=00012206
eax=1c86aae4 ebx=1bff0030 ecx=00140002 edx=00000001 edi=202834b0 esi=00000000
Process dumping was disabled, use DUMPPROC / PROCDUMP to enable it.
Executing: gcc.exe -idirafter f:/lang/os2tk45/h -Wall -Wpointer-arith -Wdeclaration-after-statement -Werror=return-type -Werror=int-to-pointer-cast -Wtype-limits -Wempty-body -Wsign-compare -Wno-unused -Wcast-align -march=i486 -mtune=i686 -std=gnu99 -fgnu89-inline -fno-strict-aliasing -Zomf -fno-math-errno -pthread -DNDEBUG -DTRIMMED -g -O3 -fomit-frame-pointer -Zdll -Zhigh-mem -o mozsqlt3.dll @G:/lang/work/mozilla/mozilla-os2.git/obj-ff/db/sqlite3/src/tmpncm_ax.list -lpthread -Zomf -Zmap -Zlinker DISABLE 1121 -Zhigh-mem -g -Zlinker option symfile=mozsqlt3.dbg -Zhigh-mem -LG:/lang/work/mozilla/mozilla-os2.git/obj-ff/dist/lib -lmozglue -lmmap -Zomf mozsqlt3.def
G:/lang/work/mozilla/mozilla-os2.git/obj-ff/db/sqlite3/src/tmpncm_ax.list:
    sqlite3.o

make[3]: *** [mozsqlt3.dll] Error -11
make[3]: Leaving directory `G:/lang/work/mozilla/mozilla-os2.git/obj-ff/db/sqlite3/src'
make[2]: *** [db/sqlite3/src/libs] Error 2
```

CRASH or TRAP!!!

**Comment by dmik** [comment]
Can you show the timestamp of your PTHR01.DLL? Something's still wrong on your system DLL-wide.

**Comment by komh** [comment]
PTHR01.DLL
```
15-05-21  22:27           6,298    124  pthr01.dll
```

PYTHON2.7
```
15-02-27   2:24           1,012    124  python2.7.exe
```

**Comment by dmik** [comment]
It looks exactly like mine. Then it's something else they refer (mmap? the lack of TMP/TEMP?). I recall that some one else had this PTRH01 crash and it was in that area but I don't remember exactly what it was. I will point @ydario at it. Or you may try to ask him yourself on the channel.

**Comment by dryeo** [comment]
It could also be hardware related. Wouldn't hurt to run memcheck and 
keep an eye on the system temperatures. The build does stress things.

**Comment by dmik** [comment]
Turned out that the new photon build system symlinks many files in `dest` instead of copying them (this relates at least to the `include` and `res` subdirs (EXPORTS and RESOURCE_FILES containers, respectively). This looks cool (saves a lot of space) but unfortunately, resource files are read by Firefox using nsIFileInputStream which eventually ends up in PR_Read (NSPR) which uses DosRead which is unaware of kLIBC symlinks. For this reason, I had to disable symlinking for `res` files in b9a97e21b02dd4136884adcf2b9eec00b20d1250. Now the `about:home` page works correctly. This commit could also fix some other cases like the ones where `css` files are read from `resource://`.

But we should really port NSPR to libc I/O one day (instead of using Dos* APIs directly) — many apps will benefit from this.

**Comment by dryeo** [comment]
Fix b9a97e2 removed about a third of SM problems. Todays build stderr and stout log is here, https://gist.github.com/dryeo/4879294e1ca6fef8dc6a.

**Comment by dmik** [comment]
Good to know that.

**Comment by dmik** [comment]
It seems that the shutdown crash is related to the new `osfile` code. There is some new async file I/O operation logic that doesn't work properly, apparently. Kinda hard to debug because of JS (in particular, because of its very weak typed nature which makes it hard to read the code if you don't know it well).

**Comment by dryeo** [comment]
It actually uses CTYPES to call native file (and I think threading) 
operations in the kernel or libc. The main bug is 
https://bugzilla.mozilla.org/show_bug.cgi?id=563742.
Note that our CTYPES port is probably broken and will need work to make 
calls using the right calling convention.

**Comment by dmik** [comment]
Sure, I know that. As I did the ctype-based JS wrappers for our libc once it was first introduced on our branch (esr24). The thing is that in esr31 they go further: these libc calls are performed asynchronously (to avoid blocking the UI, obviously). And this is where something doesn't currently work. I doubt this relates to the calling convention as it worked in esr24 (unless they did some major change in ctypes that I missed). I'm checking it all right now.

**Comment by dryeo** [comment]
OK, I didn't realize that 24ESR actually used CTYPES and it worked. When 
I did the CTYPES port, it was just a quick hack to use stdcall and I 
could not test as the testsuite depended on TCL/TK (wish.exe?) and some 
testing I did for enigmail back in the 10ESR days failed

**Comment by dmik** [comment]
Yes, actually I did quite a lot to make it work at the JS level. I almost didn't touch the `libffi` library itself though, it worked well for me out-of-the-box.

BTW, I fixed the unexpected shutdown delay (and the subsequent crash initiated by the watchdog thread that does so after a minute of waiting to avoid leaving the user with a browser which is stuck forever). Turned out that there was a merge typo which removed a few necessary lines when adopting this after it was changed and even moved to another directory upstream. See 19f00c7cc377d12a86bf7ccb7364f68b90024f6e for details. Now I get a quick shutdown when I do File->Quit and no crash. Note that this also fixed creation of many necessary files in the Profile directory (such as `sessionCheckpoints.json`, `bookmarkbackups` and many others) — because of the typo, the async worker was not functioning at all (after being terminated with a fatal error) and all I/O operations would simply stuck.

Now I basically need to sort out `Unix error 110` at startup (also in JS code) and I think we are ready for an alpha build for our internal testers.

**Comment by dmik** [comment]
I sorted it out: there was some mess with error code handling, see 8ea94f5d2f2f71706254f36894ef39302edda9b7 and a6369969d5e51ad6b0c6b5c9284e6c0e29efb115 for details. The root of the problem is that our NSPR uses Dos* APIs and returns OS/2 error codes while osfile expects Posix errno (because we follow the Posix code path there by wrapping libc calls).

The file `sessionstore.js` is now successfully created and maintained. The `extensions.json` file that couldn't be manipulated due to some weird access permission error, now works well to. This means that all issues I've initially spotted are resolved. I will try a release build now to see how it behaves.

**Comment by wztest** [comment]
Good to see that the base functions are now working after the latest fixes, sessionstore.js is one of the important things in case the browser crashes because of other issues.

**Comment by dmik** [comment]
It behaves quite well. I'm closing this issue. New issues should be reported separately.

**Comment by dryeo** [comment]
ESR31.8 has been tagged and will be released next week I believe
